xarray/core/indexing.py

```diff
@@ -837,12 +873,17 @@ class MemoryCachedArray(ExplicitlyIndexedNDArrayMixin):
     def __init__(self, array):
         self.array = _wrap_numpy_scalars(as_indexable(array))
 
-    def _ensure_cached(self):
-        self.array = as_indexable(self.array.get_duck_array())
-
     def get_duck_array(self):
-        self._ensure_cached()
-        return self.array.get_duck_array()
+        duck_array = self.array.get_duck_array()
+        # ensure the array object is cached in-memory
+        self.array = as_indexable(duck_array)
+        return duck_array
+
+    async def async_get_duck_array(self):
+        duck_array = await self.array.async_get_duck_array()
+        # ensure the array object is cached in-memory
+        self.array = as_indexable(duck_array)
+        return duck_array
```

Comment:
Might need a deep copy here to match previous behavior

Reply:
I added this in https://github.com/pydata/xarray/pull/10327/commits/da2d43cc31b454a5ae2f782fbcecd6ed4307ab56 but I'm very unsure if this is what you meant